src/shared/protocol.ts

```diff
@@ -360,31 +510,100 @@ export abstract class Protocol<SendRequestT extends Request, SendNotificationT e
         const abortController = new AbortController();
         this._requestHandlerAbortControllers.set(request.id, abortController);
 
-        const fullExtra: RequestHandlerExtra<SendRequestT, SendNotificationT> = {
+        const taskMetadata = request.params?._meta?.[TASK_META_KEY];
+        const fullExtra: RequestHandlerExtra<SendRequestT, SendNotificationT, SendResultT> = {
             signal: abortController.signal,
             sessionId: capturedTransport?.sessionId,
             _meta: request.params?._meta,
-            sendNotification: notification => this.notification(notification, { relatedRequestId: request.id }),
-            sendRequest: (r, resultSchema, options?) => this.request(r, resultSchema, { ...options, relatedRequestId: request.id }),
+            sendNotification: async notification => {
+                const relatedTask = taskMetadata ? { taskId: taskMetadata.taskId } : undefined;
+                await this.notification(notification, { relatedRequestId: request.id, relatedTask });
+            },
+            sendRequest: async (r, resultSchema, options?) => {
+                const relatedTask = taskMetadata ? { taskId: taskMetadata.taskId } : undefined;
+                if (taskMetadata && this._taskStore) {
+                    await this._taskStore.updateTaskStatus(taskMetadata.taskId, 'input_required');
+                }
+                try {
+                    return await this.request(r, resultSchema, { ...options, relatedRequestId: request.id, relatedTask });
+                } finally {
+                    if (taskMetadata && this._taskStore) {
+                        await this._taskStore.updateTaskStatus(taskMetadata.taskId, 'working');
```

Comment:
This also seems like it could race. While hopefully a request handler never completes before a request it makes, timeouts and the like make this inevitable.

If the expectation is that the TaskStore should be responsible for dealing with this race, that's all the more reason to include a production-ready in-memory TaskStore in the box that makes it clear what the expectations are for a TaskStore to be production-ready.